# UBDC Data Dives - Oct 20th, 2020
# An exploration of fuel poverty in the private rental housing market in City of Glasgow
## Dr. Qunshan Zhao, Lecturer in Urban Analytics, University of Glasgow

## Part 1: Zoopla data cleaning and manipulation

### Basic information of Zoopla data:

### Zoopla is one of the largest UK online platfrom for house sale and rental activities, you can visit their official website through https://www.zoopla.co.uk/.
### Zoopla data can be obtained through UBDC data services: 
### https://www.ubdc.ac.uk/data-services/data-catalogue/housing-data/zoopla-property-data/

### Zoopla data time range: 01/01/2010 - 31/03/2019

### Zoopla files: 188 csv, 44.3 GB

### Coverage: Great Britain (England, Wales, Scotland) 

In [29]:
import pandas as pd
import glob

### Read Zoopla data into pandas dataframe

In [30]:
path = r'D:\UBDC Data\Datasets\Zoopla Property Listings\safeguarded-release-zoopla-historical-all-uk-21032019\all_uk'
all_files = glob.glob(path + "/*.csv")

li = []

for filename in all_files:
    df = pd.read_csv(filename, index_col=None, header=0)
    li.append(df)

frame = pd.concat(li, axis=0, ignore_index=True)

In [31]:
frame.head()

listing_id  property_id ttwa11nm ttwa11cd plan_no  extract_date  \
0    45072996     10001276      NaN      NaN     NaN           NaN   
1    45878620     10004514      NaN      NaN     NaN           NaN   
2    44332648     10004933      NaN      NaN     NaN           NaN   
3    45194644     10006862      NaN      NaN     NaN           NaN   
4    45200783     10007177      NaN      NaN     NaN           NaN   

   result_count                                        details_url  \
0             1  https://www.zoopla.co.uk/property-history/the-...   
1             1  https://www.zoopla.co.uk/property-history/27-n...   
2             1  https://www.zoopla.co.uk/property-history/5/av...   
3             1  https://www.zoopla.co.uk/property-history/denm...   
4             1  https://www.zoopla.co.uk/property-history/319-...   

                                         description  \
0  An attractive four bedroom detached 1920's hou...   
1  A beautifully presented period cottage full of...   
2  The property has been finished to a high speci...   
3  The accommodation is located over two floors a...   
4  A stunning character cottage located in Brookw...   

                                   short_description  ...  \
0  An attractive four bedroom detached 1920's hou...  ...   
1  <p class="top">A beautifully presented period ...  ...   
2  A brand new one double bedroom executive style...  ...   
3  The accommodation is located over two floors a...  ...   
4  A stunning character cottage located in Brookw...  ...   

                                           image_url  \
0  https://lid.zoocdn.com/354/255/7ccf354fb6a118c...   
1  https://lid.zoocdn.com/354/255/7277f68e34cd824...   
2  https://lid.zoocdn.com/354/255/56256e3b0874b13...   
3  https://lid.zoocdn.com/354/255/1e1ad2a0edc1a70...   
4  https://lid.zoocdn.com/354/255/ca9b55ef83528e7...   

                                       thumbnail_url   image_caption  \
0  https://lid.zoocdn.com/80/60/7ccf354fb6a118c02...           Front   
1  https://lid.zoocdn.com/80/60/7277f68e34cd824f7...           Front   
2  https://lid.zoocdn.com/80/60/56256e3b0874b1382...            Main   
3  https://lid.zoocdn.com/80/60/1e1ad2a0edc1a709d...           Front   
4  https://lid.zoocdn.com/80/60/ca9b55ef83528e76f...  Picture No. 24   

                                 original_image_list  \
0  https://lc.zoocdn.com/7ccf354fb6a118c020efbaa8...   
1  https://lc.zoocdn.com/7277f68e34cd824f7935a1c9...   
2  https://lc.zoocdn.com/56256e3b0874b13828a80533...   
3  https://lc.zoocdn.com/1e1ad2a0edc1a709dac1d231...   
4  https://lc.zoocdn.com/ca9b55ef83528e76f9add9bd...   

                                    other_image_list  \
0  https://lid.zoocdn.com/354/255/e5e8dd93fa31bef...   
1  https://lid.zoocdn.com/354/255/15d1a4d9e8cc86e...   
2  https://lid.zoocdn.com/354/255/0143e8fbb192c4c...   
3  https://lid.zoocdn.com/354/255/0b6d41fdcfbd8e9...   
4  https://lid.zoocdn.com/354/255/1ab1d478d8a950e...   

                               other_image_desc_list  \
0  Lounge|Kitchen|Dining Room|Bedroom 2|Garden|Ga...   
1  Living Room|Kitchen|Garden|Dining Area|Bedroom...   
2  Living Room|Open-Plan Kitchen|Inner Hall|Doubl...   
3  Lounge|Rear Views|Lounge|Dining Room|Kitchen|B...   
4  Picture No. 22|Picture No. 04|Picture No. 05|P...   

                                       document_list  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  http://seymours.reapitcloud.com/seyrps/pdf.php...   

                                            epc_list  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3  https://lc.zoocdn.com/6d83318f27040130182dac9c...   
4                                                NaN   

  

In [32]:
frame.shape

(13132108, 52)

### Filter the data to Glasgow region

In [33]:
Glasgow = frame[frame.post_town == 'Glasgow']

In [34]:
Glasgow.shape

(124006, 52)

### Further filter the data to Glasgow City based on postcode

In [ ]:
# Filter rows based on column value in pandas:
# https://cmdlinetips.com/2018/02/how-to-subset-pandas-dataframe-based-on-values-of-a-column/
# Glasgow City postcode
# https://en.wikipedia.org/wiki/G_postcode_area
GlasgowCityPostCode = ['G1', 'G2','G3', 'G4','G5', 'G11','G12', 'G13','G14', 'G15','G20', 'G21','G22', 'G23','G31', 'G32', 'G33', 'G34','G40', 'G41','G42', 'G43','G44', 'G45','G46', 'G51','G52', 'G53']
GlasgowCity = Glasgow[Glasgow.outcode.isin(GlasgowCityPostCode)]

In [38]:
GlasgowCity.shape

(71496, 52)

### Since we are interested in the private rental market in this analysis, we only focus on rental properties. 

In [ ]:
GlasgowCityRent = GlasgowCity[GlasgowCity.listing_status == 'rent']

In [40]:
GlasgowCityRent.shape

(17647, 52)

### Now we start to do some data cleaning. First print all the column names here. 

In [42]:
print(GlasgowCityRent.columns.values)

['listing_id' 'property_id' 'ttwa11nm' 'ttwa11cd' 'plan_no' 'extract_date'
 'result_count' 'details_url' 'description' 'short_description'
 'bullet_list' 'outcode' 'incode' 'price' 'price_change'
 'price_change_percent_list' 'price_change_date_list'
 'price_change_direction_list' 'price_change_summary.percent'
 'price_change_summary.direction' 'price_change_summary.last_updated_date'
 'first_marketed_date' 'last_marketed_date' 'street_name'
 'property_number' 'property_type' 'category' 'num_bedrooms' 'num_floors'
 'num_bathrooms' 'num_recepts' 'county' 'country' 'post_town'
 'displayable_address' 'agent_address' 'agent_postcode' 'agent_name'
 'agent_logo' 'listing_status' 'status' 'branch_id' 'image_url'
 'thumbnail_url' 'image_caption' 'original_image_list' 'other_image_list'
 'other_image_desc_list' 'document_list' 'epc_list' 'virtual_tour_list'
 'floor_plan_list']


### We use last_marketed_date as the final date for each of the posting. N/A value exists in this column and we need to remove it. 

In [227]:
GlasgowCityRent.last_marketed_date.describe()

count                   17647
unique                   9043
top       0000-00-00 00:00:00
freq                     1572
Name: last_marketed_date, dtype: object

In [300]:
GlasgowCityRent_R1 = GlasgowCityRent[GlasgowCityRent.last_marketed_date != '0000-00-00 00:00:00']

In [228]:
GlasgowCityRent_R1.shape

(16075, 52)

### Convert pandas series to datetime format

In [46]:
# https://stackoverflow.com/questions/28133018/convert-pandas-series-to-datetime-in-a-dataframe
GlasgowCityRent_R1["last_marketed_date"] = pd.to_datetime(GlasgowCityRent_R1["last_marketed_date"])

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


### Analyze bedroom numbers

In [48]:
GlasgowCityRent_R1['num_bedrooms'].value_counts().sort_index()

0      143
1     4808
2     8000
3     2459
4      478
5      157
6       23
7        5
8        1
22       1
Name: num_bedrooms, dtype: int64

### I only keep bedroom numbers range from 1-5 to remove outliers. Overall I only remove slightly more than 1% of the data. 

In [49]:
Num_bedrooms = [1,2,3,4,5]
GlasgowCityRent_R2 = GlasgowCityRent_R1[GlasgowCityRent_R1.num_bedrooms.isin(Num_bedrooms)]

In [50]:
GlasgowCityRent_R2.shape

(15902, 52)

In [51]:
GlasgowCityRent_R2['num_bedrooms'].value_counts().sort_index()

1    4808
2    8000
3    2459
4     478
5     157
Name: num_bedrooms, dtype: int64

### Here we show the rental status. In the Zoopla data, we have three status including rental under offer, rented, or to be rented. I used all the data in this analysis and assumed all the rental properties has been rented out eventually. 

In [231]:
GlasgowCityRent_R2['status'].value_counts().sort_index()

rent_under_offer    1251
rented              4779
to_rent             9872
Name: status, dtype: int64

### Here I further clean the data based on time. Since the Scottish EPC data has a range from Oct 2012 - Mar 2020. I cleaned the Zoopla data to Jan 2013 - Dec 2018 to ensure the same time coverage with EPC data. 

### It is worth mentioned that I do not separate different years in this analysis, which can be done in the future analysis.

In [232]:
GlasgowCityRent_R2['last_marketed_date'].dt.year.value_counts().sort_index()

2010       1
2011      39
2012     996
2013    3016
2014    3802
2015    2571
2016    2551
2017    2220
2018     693
2019      13
Name: last_marketed_date, dtype: int64

In [233]:
Year = [2013,2014,2015,2016,2017,2018]
GlasgowCityRent_R3 = GlasgowCityRent_R2[GlasgowCityRent_R2['last_marketed_date'].dt.year.isin(Year)]

In [234]:
GlasgowCityRent_R3['last_marketed_date'].dt.year.value_counts().sort_index()

2013    3016
2014    3802
2015    2571
2016    2551
2017    2220
2018     693
Name: last_marketed_date, dtype: int64

In [235]:
GlasgowCityRent_R3.shape

(14853, 52)

### Since Zoopla provides the weekly rental price in their dataset, it makes less sense to include very large rental price since they can be either errors or outliers (huge villa or expensive houses, etc.). I set the weekly rental price < 500 and it removes top 1% outliers from the dataset. 

In [237]:
GlasgowCityRent_R3['price'].sort_values(ascending=False)

4190072     87692
4119587     69000
4153832     57692
4153712     56538
3841656     49615
3677648     47308
3812353     45692
3735830     45000
3876300     43835
4154282     43615
3703335     41308
3680114     40385
138817      38077
12401422    36692
3703797     34154
6405654     32308
4049207     29999
4084881     28846
4734112     28845
3735985     27692
4088215     26538
3734481     25385
3878914     25154
4193016     24692
4120007     24231
7440763     23065
3677488     21923
3734271     21923
3805002     21450
3676391     19615
            ...  
3767604        65
3909164        65
3836786        63
3836788        63
3836787        63
4118337        62
4189087        61
3768031        60
3768020        60
3768028        60
3768030        60
3877207        58
6947060        58
3836785        58
4015984        58
3736184        58
3911130        57
7645534        57
6405649        55
6405648        55
3739453        53
7012380        46
4048122        46
4116503        46
6947056   

In [301]:
GlasgowCityRent_R4 = GlasgowCityRent_R3[GlasgowCityRent_R3.price < 500]

In [302]:
GlasgowCityRent_R4.shape

(14711, 52)

### Zoopla data manipulation

### Calculate yearly rent by price * 52 weeks

In [190]:
GlasgowCityRent_R4['YearlyRent'] = GlasgowCityRent_R4['price']*52

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


### Generate full postcode column

In [242]:
GlasgowCityRent_R4['FullPostcode'] = GlasgowCityRent_R4['outcode'] + ' ' + GlasgowCityRent_R4['incode']

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


### Geerante postcode sector column to match with EPC data

In [243]:
GlasgowCityRent_R4['Postcode Sector'] = GlasgowCityRent_R4['FullPostcode'].astype(str).str[:-2]

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


### A statistics of zoopla listings in each of the postcode sector, range from 7 listings to 639 listings. 

In [247]:
GlasgowCityRent_R4['Postcode Sector'].value_counts().sort_values()

G15 7      7
G40 3      8
G2 2       8
G34 9      9
G2 1      10
G22 7     12
G15 8     12
G2 6      13
G33 4     16
G15 6     22
G2 7      23
G21 2     24
G34 0     27
G1 2      30
G52 4     32
G33 3     34
G22 5     34
G1 3      35
G33 5     42
G45 0     44
G40 4     45
G42 0     46
G46 6     48
G13 4     49
G45 9     50
G2 3      51
G51 3     52
G32 6     53
G5 9      54
G21 3     54
        ... 
G31 3    176
G5 8     178
G4 0     180
G52 2    184
G53 7    195
G20 8    200
G40 1    216
G1 5     227
G44 5    247
G41 2    256
G20 6    278
G3 7     279
G14 0    284
G1 1     287
G31 2    294
G13 1    306
G12 0    312
G44 4    320
G11 6    327
G4 9     353
G42 9    353
G11 5    360
G12 8    360
G3 6     368
G51 1    395
G3 8     416
G11 7    458
G42 8    495
G12 9    564
G41 3    639
Name: Postcode Sector, Length: 100, dtype: int64

### An overview of property type in the dataset

In [248]:
GlasgowCityRent_R4['property_type'].value_counts().sort_index()

Bungalow                     29
Chalet                        1
Cottage                     124
Detached bungalow            12
Detached house              188
End terrace house            85
Finca                         1
Flat                      13060
Maisonette                   46
Mews house                   23
Semi-detached bungalow       16
Semi-detached house         392
Studio                       56
Terraced house              404
Town house                   33
Villa                        22
Name: property_type, dtype: int64

### Calculate the mean value for yearly rental cost in each of the postcode sector

In [193]:
GlasgowCityRent_R5 = GlasgowCityRent_R4.groupby(['Postcode Sector']).mean()

In [249]:
GlasgowCityRent_R5['YearlyRent'].sort_values()

Postcode Sector
G34 0     4982.370370
G40 3     5083.000000
G31 5     5085.459459
G32 7     5337.382716
G34 9     5379.111111
G33 4     5534.750000
G42 7     5594.216216
G42 8     5687.749495
G51 2     5688.179104
G31 3     5714.681818
G31 4     5744.631579
G40 4     5754.666667
G45 9     5809.440000
G53 6     5821.214286
G14 0     5885.154930
G21 4     5892.455696
G21 2     5902.000000
G21 1     5902.590909
G32 8     5915.000000
G21 3     5954.962963
G44 4     5977.075000
G32 0     6004.227273
G40 1     6030.555556
G51 3     6034.000000
G5 9      6034.888889
G44 5     6059.368421
G42 0     6061.391304
G40 2     6151.407407
G23 5     6204.571429
G15 7     6277.142857
             ...     
G5 8      8163.707865
G1 2      8183.066667
G13 1     8340.052288
G4 0      8366.800000
G33 1     8413.600000
G44 3     8532.588235
G1 5      8611.841410
G41 5     8729.573034
G20 6     8774.345324
G3 8      8908.375000
G11 6     8933.504587
G1 1      9123.191638
G2 8      9158.666667
G11 5     9288.9

## Part 2: EPC data cleaning and manipulation

### Scottish government shares Scottish EPC records through Energy Saving Trust.

### Note that not all EPCs are available on the register as central Lodgement was phased in. EPC data available is as follows:

### All buildings which were newly constructed, with a building warrant application made after 9 January 2013
### Dwellings sold or rented to a new tenant after 1 December 2008; and
### Non-domestic buildings sold, rented to a new tenant and public buildings from January 2013. https://www.scottishepcregister.org.uk/

### Individual EPC data can be found and downloaded through the website, but the Scottish government only makes the EPC data available at Postcode Sectors level and can be downloaded from here: https://statistics.gov.scot/data/domestic-energy-performance-certificates

### EPC data: 2.5GB, 30 csv. 

### EPC data time range: Oct 2012 - Mar 2020. 

### Coverage: Scotland

### Read EPC data into pandas data frame

In [67]:
path = r'D:\SEPCR - EPC Extract May 2020 - Published - D EPC data extract 06 2020' 
all_files = glob.glob(path + "/*.csv")

epc = []

for filename in all_files:
    df = pd.read_csv(filename, index_col=None, header=0)
    epc.append(df)

epcframe = pd.concat(epc, axis=0, ignore_index=True)

In [68]:
epcframe.shape

(1271423, 49)

### I used date of assessment as the EPC obtained date. The code actually found out year errors in the EPC data, it would be good to report to the Scottish government for correction.

In [69]:
# https://stackoverflow.com/questions/32888124/pandas-out-of-bounds-nanosecond-timestamp-after-offset-rollforward-plus-adding-a
epcframe["Date of Assessment"] = pd.to_datetime(epcframe["Date of Assessment"], errors = 'coerce')

In [70]:
epcframe['Date of Assessment'].dt.year.value_counts().sort_index()

1970.0         1
2009.0        42
2010.0        45
2011.0       666
2012.0     14980
2013.0    125247
2014.0    194799
2015.0    174584
2016.0    170576
2017.0    167471
2018.0    177018
2019.0    210733
2020.0     35170
2103.0         1
2108.0         1
Name: Date of Assessment, dtype: int64

### Same for Zoopla data, we further filter the EPC data range to 2013-2018.

In [71]:
Year = [2013,2014,2015,2016,2017,2018]
EPC = epcframe[epcframe['Date of Assessment'].dt.year.isin(Year)]

In [72]:
EPC.shape

(1009695, 49)

### Create postcode district column

In [253]:
EPC['Postcode'] = EPC['Postcode Sector'].astype(str).str[:-2]

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


#### Extract Glasgow data based on postcode district

In [255]:
GlasgowCityPostCode = ['G1', 'G2','G3', 'G4','G5', 'G11','G12', 'G13','G14', 'G15','G20', 'G21','G22', 'G23','G31', 'G32', 'G33', 'G34','G40', 'G41','G42', 'G43','G44', 'G45','G46', 'G51','G52', 'G53']
EPC_R1 = EPC[EPC.Postcode.isin(GlasgowCityPostCode)]

In [256]:
EPC_R1.shape

(128873, 50)

### I calculate the energy cost per year per square meter for future use.

In [257]:
EPC_R1['Unit Energy cost per year'] = EPC_R1['Total current energy costs over 3 years (£)']/(EPC_R1['Total floor area (m²)']*3)

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


### Here are the numbers of EPC records in each of the postcode sector, range from 1 to 2660. 

In [259]:
EPC_R1['Postcode Sector'].value_counts().sort_values()

G20 5       1
G21 6       1
G32 5       1
G22 8       6
G2 6       21
G2 5       23
G2 1       48
G2 2       88
G1 3      103
G2 7      107
G2 4      138
G1 2      157
G2 8      212
G1 4      253
G2 3      298
G5 8      442
G40 4     504
G5 9      560
G40 3     591
G41 1     627
G52 4     696
G40 2     734
G41 4     736
G31 1     750
G1 5      786
G34 0     794
G41 5     837
G1 1      875
G15 8     884
G22 5     919
         ... 
G22 7    1602
G41 2    1603
G45 9    1610
G43 2    1615
G11 5    1674
G12 0    1723
G31 4    1725
G13 1    1727
G46 8    1754
G32 7    1762
G14 9    1793
G52 2    1843
G31 2    1886
G43 1    1888
G32 6    1890
G44 4    1946
G13 3    1988
G21 4    1995
G22 6    2004
G12 9    2125
G14 0    2191
G5 0     2201
G21 3    2227
G3 8     2290
G42 9    2296
G31 3    2314
G11 7    2408
G42 8    2451
G41 3    2620
G53 7    2660
Name: Postcode Sector, Length: 105, dtype: int64

### Calculate the mean of unit energy cost in each postcode sector.

In [260]:
EPC_R2 = EPC_R1.groupby(['Postcode Sector']).mean()

In [262]:
EPC_R2['Unit Energy cost per year'].sort_values()

Postcode Sector
G2 6      6.927343
G5 0      7.047091
G22 8     7.448117
G5 9      8.227051
G42 0     8.407751
G53 7     8.438179
G45 0     8.592793
G41 1     8.670550
G15 8     8.766024
G45 9     8.784711
G33 4     8.786965
G1 3      8.839071
G1 4      8.839481
G3 8      8.858424
G11 6     8.922749
G32 6     8.998660
G20 9     9.060933
G21 6     9.111111
G31 5     9.129714
G31 4     9.140985
G34 9     9.218425
G22 5     9.283547
G5 8      9.326112
G21 3     9.509951
G33 1     9.525798
G15 7     9.574503
G33 3     9.576014
G14 0     9.592844
G40 4     9.658710
G33 6     9.689953
           ...    
G1 5     11.257643
G32 9    11.286636
G52 1    11.311989
G46 7    11.314480
G12 8    11.318985
G44 5    11.320098
G32 0    11.366514
G34 0    11.468580
G12 9    11.469589
G42 7    11.504992
G52 3    11.514105
G4 9     11.526398
G11 7    11.558193
G44 3    11.570301
G42 9    11.599356
G13 3    11.621626
G20 6    11.622309
G4 0     11.772666
G42 8    11.817110
G52 4    11.882820
G51 1    11.973

### Now I finished the data cleaning and manipulation of the Zoopla and EPC data. The next step is to join these two datasets together based on Postcode sector

In [263]:
Zoopla_EPC = GlasgowCityRent_R5.merge(EPC_R2, left_on='Postcode Sector', right_on='Postcode Sector')

### After joining these two dataset, I calculated the energy cost per year based on the bedroom numbers from Zoopla data and unit energy cost from EPC data in each postcode sector.

In [265]:
# We don't separate different years in this analysis. 
Zoopla_EPC['num_bedrooms'].describe()

count    100.000000
mean       2.031444
std        0.329729
min        1.333333
25%        1.801685
50%        2.042278
75%        2.209967
max        2.978947
Name: num_bedrooms, dtype: float64

### I used 35 m2 for one bedroom, 70 m2 for two bedrooms, and 105 m2 for three bedrooms in the analysis based on the reference below. This can be further improved based on other resources.
### Space standard for homes: Under the new standard, a new one bed, one person flat would have to be a minimum of 37m² while a three bed, five person home would be a minimum of 93m².

### Source: https://hqnetwork.co.uk/download.cfm?doc=docm93jijm4n2909.pdf&ver=5651. 

In [266]:
Zoopla_EPC['EnergyCostPerYear'] = Zoopla_EPC['num_bedrooms']*35*Zoopla_EPC['Unit Energy cost per year']

### Total energy and rental cost in each postcode sector

In [267]:
Zoopla_EPC['TotalEnerngyRentalCost'] = Zoopla_EPC['EnergyCostPerYear'] + Zoopla_EPC['YearlyRent']

In [269]:
Zoopla_EPC['TotalEnerngyRentalCost'].sort_values()

Postcode Sector
G31 5     5612.268649
G40 3     5669.704887
G34 0     5889.237686
G32 7     5940.327684
G34 9     6096.099721
G33 4     6226.723501
G42 7     6239.039924
G31 3     6296.554273
G51 2     6311.929430
G31 4     6317.145923
G42 8     6366.218740
G40 4     6385.702389
G14 0     6440.796787
G5 9      6461.476712
G45 9     6510.459970
G53 6     6555.759880
G21 2     6601.799343
G21 1     6616.391252
G40 1     6630.507383
G21 4     6632.329142
G32 8     6634.995727
G21 3     6700.789641
G32 0     6732.071674
G42 0     6739.494705
G44 4     6750.626373
G51 3     6800.458723
G40 2     6832.609287
G23 5     6911.798780
G51 1     6974.252287
G44 5     7007.369336
             ...     
G53 7     8787.566453
G4 0      9081.008407
G13 1     9147.224996
G15 6     9192.923032
G1 5      9306.145392
G33 1     9383.033114
G44 3     9425.883566
G3 8      9474.803084
G11 6     9475.009023
G20 6     9508.891998
G1 1      9713.215525
G2 8      9727.267531
G41 5     9738.387568
G2 6     10002.3

### Now I have finished the data analysis for Zoopla and EPC data. To understand the spatial distribution of fuel poverty, I will need income information and postcode shapefiles. Geopandas has been used here to deal with spatial dataset. 

In [204]:
import geopandas as gpd

#### Read postcode sector shapefile from National Records of Scotland: https://www.nrscotland.gov.uk/statistics-and-data/geography/our-products/scottish-postcode-directory/2020-2

In [270]:
PostcodeShapefile = gpd.read_file('D:\Sector_20_2\Sector_20_2.shp')

### Filter the data to Glasgow City area

In [271]:
PostcodeShapefile['District'] = PostcodeShapefile['Sector'].astype(str).str[:-2]

In [272]:
PostcodeShapefile.head()

OBJECTID  Sector    Shape_Leng    Shape_Area  \
0         1  AB10 1   7162.481830  7.162994e+05   
1         2  AB10 6  12381.783121  1.504445e+06   
2         3  AB10 7  11425.440515  2.372945e+06   
3         4  AB11 5  13779.103666  1.219521e+06   
4         5  AB11 6   7082.176177  7.724311e+05   

                                            geometry District  
0  POLYGON ((394229.854 806675.589, 394239.046 80...     AB10  
1  POLYGON ((393348.999 805786.000, 393364.000 80...     AB10  
2  POLYGON ((393209.011 805132.125, 393206.363 80...     AB10  
3  POLYGON ((395211.917 806599.168, 395275.000 80...     AB11  
4  MULTIPOLYGON (((393211.114 805537.072, 393207....     AB11

In [208]:
GlasgowCityPostCode = ['G1', 'G2','G3', 'G4','G5', 'G11','G12', 'G13','G14', 'G15','G20', 'G21','G22', 'G23','G31', 'G32', 'G33', 'G34','G40', 'G41','G42', 'G43','G44', 'G45','G46', 'G51','G52', 'G53']
PostcodeGlasgow = PostcodeShapefile[PostcodeShapefile.District.isin(GlasgowCityPostCode)]

### Merge Zoopla and EPC data with the shapefile

In [209]:
PostcodeGlasgow_J1 = PostcodeGlasgow.merge(Zoopla_EPC, left_on = 'Sector', right_on = 'Postcode Sector')

In [210]:
PostcodeGlasgow_J1.head()

OBJECTID Sector   Shape_Leng     Shape_Area  \
0       346   G1 1  4541.467998  334875.989304   
1       347   G1 2  4990.616669  184045.902873   
2       348   G1 3  1969.103068  103372.740251   
3       349   G1 4  3284.953503  195767.064752   
4       350   G1 5  4034.758081  437864.310824   

                                            geometry District    listing_id  \
0  POLYGON ((259437.000 665598.000, 259482.000 66...       G1  3.481988e+07   
1  MULTIPOLYGON (((258979.851 665753.005, 259004....       G1  3.856798e+07   
2  POLYGON ((258831.731 665351.411, 258837.117 66...       G1  3.747108e+07   
3  MULTIPOLYGON (((258309.000 664914.950, 258294....       G1  3.666225e+07   
4  POLYGON ((259449.111 665128.304, 259445.088 66...       G1  3.571428e+07   

    property_id  extract_date  result_count  ...  \
0  1.971486e+07  2.017070e+13      8.912892  ...   
1  1.769373e+07  2.017064e+13      3.233333  ...   
2  2.362394e+07  2.017068e+13      6.000000  ...   
3  2.243601e+07  2.017067e+13      8.325581  ...   
4  2.041333e+07  2.017069e+13      5.621145  ...   

   Potential future savings over 3 years (£).1  Space Heating  Water Heating  \
0                                   651.774857    4148.216000    2027.432000   
1                                   765.936306    4140.662420    1937.318471   
2                                   545.854369    4371.524272    2046.174757   
3                                   280.173913    3326.478261    2090.826087   
4                                   446.709924    4346.835878    1910.590331   

   Impact Of Loft Insulation  Impact Of Cavity Wall Insulation  \
0                -159.780571                        -14.701714   
1                -122.095541                       -112.057325   
2                -152.485437                          0.000000   
3                -120.249012                         -7.494071   
4                -352.998728                       -159.400763   

   Impact Of Solid Wall Insulation  Addendum Text  Unit Energy cost per year  \
0                      -456.217143            NaN                 101.029612   
1                      -131.210191            NaN                 114.455961   
2                      -504.679612            NaN                  79.551640   
3                      -383.778656            NaN                  79.555328   
4                      -830.279898            NaN                 101.318788   

   EnergyCostPerYear  TotalEnerngyRentalCost  
0        5310.214983            14433.406621  
1        5341.278167            13524.344834  
2        4773.098422            15721.326994  
3        5698.381646            15840.195600  
4        6248.735839            14860.577249  

[5 rows x 45 columns]

### Read the weekly income data from Scottish government website: https://statistics.gov.scot/data/local-level-average-household-income-estimates-2014

In [273]:
Income = pd.read_csv('D:\Local Level Average Household Income Estimates 2014.csv')

In [279]:
Income['Measurement'].value_counts()

Mean      6970
Median    6970
Name: Measurement, dtype: int64

### I removed median value in the income data and used mean income in this analysis.

In [281]:
Income_R1 = Income[Income.Measurement != 'Median']

### Since the income data is at datazone level, I read another postcode (full postcode) shapefile from National Records of Scotland: https://www.nrscotland.gov.uk/statistics-and-data/geography/our-products/scottish-postcode-directory/2020-2

In [282]:
FullPostcode = gpd.read_file('D:\PC_Cut_20_2\PC_Cut_20_2.shp')

### Filter the data to Glasgow City

In [216]:
GlasgowCityPostCode = ['G1', 'G2','G3', 'G4','G5', 'G11','G12', 'G13','G14', 'G15','G20', 'G21','G22', 'G23','G31', 'G32', 'G33', 'G34','G40', 'G41','G42', 'G43','G44', 'G45','G46', 'G51','G52', 'G53']
FullPostcodeGlasgow = FullPostcode[FullPostcode.District.isin(GlasgowCityPostCode)]

### This is a very useful spatial data to help convert between postcode, postcode sector, postcode district, and data zone, etc.

In [217]:
FullPostcodeGlasgow.head()

OBJECTID_1 Postcode District Sector     DoIntro    Council       OA11  \
62556       62557   G1 1AB       G1   G1 1  2018-01-19  S12000049  S00116298   
62557       62558   G1 1AD       G1   G1 1  2020-04-29  S12000049  S00112179   
62558       62559   G1 1BL       G1   G1 1  1985-09-01  S12000049  S00112170   
62559       62560   G1 1BQ       G1   G1 1  2008-04-10  S12000049  S00112177   
62560       62561   G1 1DA       G1   G1 1  1973-08-01  S12000049  S00112164   

            DZ11       EW19  Shape_Leng   Shape_Area  \
62556  S01010260  S13002976  106.398986   707.853996   
62557  S01010265  S13002976   69.183403   293.396348   
62558  S01010263  S13002976  208.320854  2593.810945   
62559  S01010265  S13002976  101.064285   484.789528   
62560  S01010265  S13002976  558.111657  5783.362461   

                                                geometry  
62556  POLYGON ((259641.880 665354.026, 259636.843 66...  
62557  POLYGON ((259174.015 665173.045, 259168.985 66...  
62558  POLYGON ((259732.886 665077.898, 259724.509 66...  
62559  POLYGON ((259324.613 665177.007, 259345.016 66...  
62560  POLYGON ((259230.553 665269.734, 259234.000 66...

### Merge weekly income data with the shapefile based on the data zone.

In [286]:
FullPostcodeGlasgow_J1 = FullPostcodeGlasgow.merge(Income_R1, left_on='DZ11', right_on='FeatureCode')

In [287]:
FullPostcodeGlasgow_J1.head()

OBJECTID_1 Postcode District Sector     DoIntro    Council       OA11  \
0       62557   G1 1AB       G1   G1 1  2018-01-19  S12000049  S00116298   
1       62573   G1 1EX       G1   G1 1  1973-08-01  S12000049  S00116299   
2       62581   G1 1HL       G1   G1 1  1973-08-01  S12000049  S00116278   
3       62582   G1 1HP       G1   G1 1  1993-09-01  S12000049  S00116278   
4       62583   G1 1JF       G1   G1 1  1973-08-01  S12000049  S00116345   

        DZ11       EW19  Shape_Leng    Shape_Area  \
0  S01010260  S13002976  106.398986    707.853996   
1  S01010260  S13002976  801.224222  16635.837712   
2  S01010260  S13002976  567.398744   8053.921991   
3  S01010260  S13002976  251.132116   2309.107658   
4  S01010260  S13002976  197.083472    992.616288   

                                            geometry FeatureCode  DateCode  \
0  POLYGON ((259641.880 665354.026, 259636.843 66...   S01010260      2014   
1  POLYGON ((259807.000 665312.000, 259800.000 66...   S01010260      2014   
2  POLYGON ((259541.000 665288.000, 259540.000 66...   S01010260      2014   
3  POLYGON ((259424.158 665297.313, 259419.418 66...   S01010260      2014   
4  POLYGON ((259454.204 665459.632, 259446.520 66...   S01010260      2014   

  Measurement         Units   Value  
0        Mean  Pounds (GBP)  460.96  
1        Mean  Pounds (GBP)  460.96  
2        Mean  Pounds (GBP)  460.96  
3        Mean  Pounds (GBP)  460.96  
4        Mean  Pounds (GBP)  460.96

### Calculate the mean weekly income for each postcode sector

In [288]:
IncomePostcodeSector = FullPostcodeGlasgow_J1.groupby(['Sector']).mean()

In [292]:
IncomePostcodeSector['Value'].sort_values()

Sector
G31 1    425.401023
G2 7     428.899630
G21 2    435.264463
G40 4    440.917312
G31 4    444.719615
G22 5    445.709182
G31 3    450.043116
G51 3    450.694437
G40 3    452.523608
G31 5    457.127576
G4 0     457.840000
G42 7    459.548293
G21 1    473.620594
G51 2    476.875952
G21 4    477.062774
G22 6    478.221509
G42 0    478.932692
G32 7    484.315440
G5 0     485.812686
G40 2    486.952909
G1 1     488.845368
G15 7    493.798033
G1 5     496.680333
G32 6    497.346980
G45 9    498.534219
G40 1    499.764267
G33 4    506.996532
G5 9     507.390694
G34 0    508.064000
G33 3    513.217299
            ...    
G3 6     599.742835
G3 8     600.976364
G32 0    610.457609
G15 6    610.508909
G23 5    618.345342
G44 4    640.654339
G20 6    645.927724
G42 9    650.781844
G11 5    653.444609
G5 8     663.902432
G41 2    667.866803
G11 7    670.349856
G52 4    670.772190
G3 7     676.641111
G41 3    681.359406
G12 8    682.522625
G44 5    690.772328
G41 5    696.097738
G14 9    716.

### Now we reach the final stage! Merge Zoopla, EPC, and income data together in a shapefile.

In [293]:
FinalGlasgow = PostcodeGlasgow_J1.merge(IncomePostcodeSector, left_on = 'Sector', right_on = 'Sector')

In [295]:
FinalGlasgow.head()

OBJECTID Sector  Shape_Leng_x   Shape_Area_x  \
0       346   G1 1   4541.467998  334875.989304   
1       347   G1 2   4990.616669  184045.902873   
2       348   G1 3   1969.103068  103372.740251   
3       349   G1 4   3284.953503  195767.064752   
4       350   G1 5   4034.758081  437864.310824   

                                            geometry District    listing_id  \
0  POLYGON ((259437.000 665598.000, 259482.000 66...       G1  3.481988e+07   
1  MULTIPOLYGON (((258979.851 665753.005, 259004....       G1  3.856798e+07   
2  POLYGON ((258831.731 665351.411, 258837.117 66...       G1  3.747108e+07   
3  MULTIPOLYGON (((258309.000 664914.950, 258294....       G1  3.666225e+07   
4  POLYGON ((259449.111 665128.304, 259445.088 66...       G1  3.571428e+07   

    property_id  extract_date  result_count  ...  \
0  1.971486e+07  2.017070e+13      8.912892  ...   
1  1.769373e+07  2.017064e+13      3.233333  ...   
2  2.362394e+07  2.017068e+13      6.000000  ...   
3  2.243601e+07  2.017067e+13      8.325581  ...   
4  2.041333e+07  2.017069e+13      5.621145  ...   

   Impact Of Solid Wall Insulation  Addendum Text  Unit Energy cost per year  \
0                      -456.217143            NaN                 101.029612   
1                      -131.210191            NaN                 114.455961   
2                      -504.679612            NaN                  79.551640   
3                      -383.778656            NaN                  79.555328   
4                      -830.279898            NaN                 101.318788   

   EnergyCostPerYear  TotalEnerngyRentalCost  OBJECTID_1  Shape_Leng_y  \
0        5310.214983            14433.406621     62604.0    281.854968   
1        5341.278167            13524.344834     62675.0    254.022975   
2        4773.098422            15721.326994     62723.5    220.219802   
3        5698.381646            15840.195600     62773.5    279.440606   
4        6248.735839            14860.577249     62843.5    284.995257   

   Shape_Area_y  DateCode       Value  
0   3525.010414    2014.0  488.845368  
1   3915.870274    2014.0  513.903617  
2   2067.454805    2014.0  546.453400  
3   3915.341295    2014.0  584.325600  
4   4865.159009    2014.0  496.680333  

[5 rows x 50 columns]

### Now we can calculate the ratio between energy cost vs income, rental cost vs income, and energy + rental cost vs income and identify fuel poverty areas. If private rental residents have spend a large portion of money on their rental and energy cost, that means they do not have much money to spend on their own life. 

In [296]:
FinalGlasgow['AllRatio'] = FinalGlasgow['TotalEnerngyRentalCost']/(FinalGlasgow['Value']*52)
FinalGlasgow['EnergyRatio'] = FinalGlasgow['EnergyCostPerYear']/(FinalGlasgow['Value']*52)
FinalGlasgow['RentalRatio'] = FinalGlasgow['YearlyRent']/(FinalGlasgow['Value']*52)

### Final dataframe and I exported the data from geopandas dataframe to shapefile for visualization

In [298]:
FinalGlasgow.head()

OBJECTID Sector  Shape_Leng_x   Shape_Area_x  \
0       346   G1 1   4541.467998  334875.989304   
1       347   G1 2   4990.616669  184045.902873   
2       348   G1 3   1969.103068  103372.740251   
3       349   G1 4   3284.953503  195767.064752   
4       350   G1 5   4034.758081  437864.310824   

                                            geometry District    listing_id  \
0  POLYGON ((259437.000 665598.000, 259482.000 66...       G1  3.481988e+07   
1  MULTIPOLYGON (((258979.851 665753.005, 259004....       G1  3.856798e+07   
2  POLYGON ((258831.731 665351.411, 258837.117 66...       G1  3.747108e+07   
3  MULTIPOLYGON (((258309.000 664914.950, 258294....       G1  3.666225e+07   
4  POLYGON ((259449.111 665128.304, 259445.088 66...       G1  3.571428e+07   

    property_id  extract_date  result_count  ...  EnergyCostPerYear  \
0  1.971486e+07  2.017070e+13      8.912892  ...        5310.214983   
1  1.769373e+07  2.017064e+13      3.233333  ...        5341.278167   
2  2.362394e+07  2.017068e+13      6.000000  ...        4773.098422   
3  2.243601e+07  2.017067e+13      8.325581  ...        5698.381646   
4  2.041333e+07  2.017069e+13      5.621145  ...        6248.735839   

   TotalEnerngyRentalCost  OBJECTID_1  Shape_Leng_y  Shape_Area_y  DateCode  \
0            14433.406621     62604.0    281.854968   3525.010414    2014.0   
1            13524.344834     62675.0    254.022975   3915.870274    2014.0   
2            15721.326994     62723.5    220.219802   2067.454805    2014.0   
3            15840.195600     62773.5    279.440606   3915.341295    2014.0   
4            14860.577249     62843.5    284.995257   4865.159009    2014.0   

        Value  AllRatio  EnergyRatio  RentalRatio  
0  488.845368  0.567798     0.208899     0.358899  
1  513.903617  0.506094     0.199876     0.306218  
2  546.453400  0.553264     0.167975     0.385290  
3  584.325600  0.521317     0.187540     0.333778  
4  496.680333  0.575381     0.241942     0.333438  

[5 rows x 53 columns]

In [299]:
FinalGlasgow.to_file("FuelPoverty.shp")